design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,143 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce new attribute `test_case`.
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Arguments
+
+```cairo
+#[test_case(1, 3)]
+#[test_case(3, 5)]
+fn my_test(a: felt252, b: u32) {
```

Comment:
Rusts' `test_case` also has a `name` parameter for each case. I think it would be good to have it here because errors would be more descriptive (the user would know which case failed)

Reply:
It'd have to be investigated in code, but we could support some variant of these:

1. `#[test_case(1, 2)]` and `#[test_case('my_name', 1, 2)]`
2. `#[test_case(1, 2)]` and `#[test_case('my_name', (1, 2))]`
3. `#[test_case(1, 2)]`and `#[test_case(case_name: 'my_name', 1, 2)]`
4. `#[test_case(1, 2)]`and `#[test_case(case_name: 'my_name', (1, 2))]`
5. `#[test_case(1, 2)]`and `#[test_case(case_name: 'my_name', arguments: (1, 2))]`

I think option 3. or 4. seem clearest and easiest to handle. What do you think

Reply:
I like the way the `test_case` crate handles it
`#[test_case(1, 2; 'my_name')]`

Arguments are the most important ones here and we do not provide unnecessary code

Reply:
I'm not sure it can be exactly replicated in cairo, but if we used `,` to separate all arguments it could be done I think. So `#[test_case(1, 2, 'my_name')]`

Reply:
That will be misleading and could be taken as a parameter. Can't we do anything in the attributes?

Reply:
Tried using `;` and it doesn't seem to be allowed, though it may require further investigation.

Reply:
maybe
`#[test_case::some_name(...)]` would be possible? Wdyt?